### ISSUE TITLE
Are targets lengths optional or not?

### ISSUE THREAD

**Issue by trishankatdatadog** [theupdateframework/specification#139]
In [Section 5.6.2 of the current spec](https://github.com/theupdateframework/specification/blob/56ef9545ea244d9c5397bce72278eb01483888d2/tuf-spec.md#5-detailed-workflows), we say:

> 5.6.2. Otherwise, download the target (up to the number of bytes specified in the targets metadata), and verify that its hashes match the targets metadata. (We download up to this number of bytes, because in some cases, the exact number is unknown. This may happen, for example, if an external program is used to compute the root hash of a tree of targets files, and this program does not provide the total size of all of these files.)

But in [Section 4.5 of the same spec](https://github.com/theupdateframework/specification/blob/56ef9545ea244d9c5397bce72278eb01483888d2/tuf-spec.md#4-document-formats), we say:

>    TARGETS is an object whose format is the following:

       { TARGETPATH : {
             "length" : LENGTH,
             "hashes" : HASHES,
             ("custom" : CUSTOM) }
         , ...
       }

So, are targets lengths optional or not?

**Comment by lukpueh** [comment]
Which of the two quotes paragraphs suggests optionality?

**Comment by trishankatdatadog** [comment]
> Which of the two quotes paragraphs suggests optionality?

The first one. However, the second one says "custom" is optional, but not "length".

**Comment by lukpueh** [comment]
But the first one says
> ... download the target (up to the **number of bytes specified in the targets metadata**), and verify that its hashes match the targets metadata. (We **download up to this number of bytes**

To me that does not suggest optionality.

**Comment by lukpueh** [comment]
But I don't really understand what an external program that is used to compute the root hash of a tree of targets files has to do with anything. Do you remember why that was added to the specification, @trishankatdatadog?

**Comment by trishankatdatadog** [comment]
> To me that does not suggest optionality.

The optionality is in "because in some cases, the exact number is unknown."

> Do you remember why that was added to the specification?

IIRC it was to allow using Merkle trees for targets files themselves. Cc @erickt @patrickvacek

**Comment by lukpueh** [comment]
> > To me that does not suggest optionality.
> 
> The optionality is in "because in some cases, the exact number is unknown."

That does not make the length in targets optional, it just says that the number might not be exact, it does not say that the number might not be there.

> 
> > Do you remember why that was added to the specification?
> 
> IIRC it was to allow using Merkle trees for targets files themselves. Cc @erickt @patrickvacek

I suggest we remove it then, because nowhere else in the specification do we mention Merkle trees (yet).

**Comment by trishankatdatadog** [comment]
Interestingly, if you look at the reference client, it requires targets lengths to be exact, not just an upper bound:

[1](https://github.com/theupdateframework/tuf/blob/771592a3425ef6ea22eff0235f119879729e7b52/tuf/client/updater.py#L3180) -> [2](https://github.com/theupdateframework/tuf/blob/771592a3425ef6ea22eff0235f119879729e7b52/tuf/client/updater.py#L1314) -> [3](https://github.com/theupdateframework/tuf/blob/771592a3425ef6ea22eff0235f119879729e7b52/tuf/download.py#L108)

**Comment by lukpueh** [comment]
That does not surprise me, just like the specification (except for the confusing note discussed above), the reference implementation client does not implement any "merkle tree root hash for target hash" logic.

**Comment by trishankatdatadog** [comment]
The surprise is in the mismatch between spec and client. Now we need to decide which one to change...

**Comment by lukpueh** [comment]
I'm not so worried about "up to number of bytes" vs. "exact number of bytes".  The purpose of the length check is to prevent downloading too much data, for which both `<` and `==`  seem equally suited.  That said, I am not opposed to aligning spec and reference implementation there. If we do so we should also check snapshot and targets download. In the spec we also say "up to number of bytes" for these, but we probably do "exact number of bytes" in the reference implementation.

Regardless, we should remove the trailing two sentences in that paragraph, because there is no reason to confuse he reader with irrelevant use cases.
> ~~(We download up to this number of bytes, because in some
cases, the exact number is unknown. This may happen, for example, if an
external program is used to compute the root hash of a tree of targets
files, and this program does not provide the total size of all of these
files.)~~

**Comment by trishankatdatadog** [comment]
> In the spec we also say "up to number of bytes" for these, but we probably do "exact number of bytes" in the reference implementation.

Precisely part of the reason for raising this issue. For all metadata files, we may never have the exact size, so we need the `<=`. For target files, we need to be clear about whether it is always `==`, or `<=` is allowed.

**Comment by erickt** [comment]
We need exact target lengths in bytes for Fuchsia. We use a content addressed file system on Fuchsia called `blobfs`, where each file (we call them `blob`s at this point) is tracked by a merkle root, and a package, which we call a `meta.far` really is just another blob that's happens to be an archive that contains a list of paths to merkle roots (plus some other stuff that's not important for this discussion). blobfs is a write-once file system for security purposes. Since we only have the roots and not the full tree, blobfs needs to know the exact size in order to tell when to compute the tree to see if we fetched it correctly. In a `meta.far`, we list the exact blob size for each blob in the package (*), but we still need to know the size of the `meta.far` in order to write it correctly into blobfs. So we use the TUF target length for this.

So if the target length isn't exact, I suppose we could add the proper length in the custom field, but that seems a little redundant.

I don't quite understand though the argument for this value to not be exact. As I read 5.6.2, it seem the use case is for targets that aren't actually pointing at a file, but instead computed over a set of other target files. If this is the case, then that might be more appropriately stored as a separate field in a targets role, rather than trying to overload the target notion.

(*): this is a bit of a lie, we currently get the blob size from the server content-length field, but we're going to be fixing this soon.

**Comment by trishankatdatadog** [comment]
Thanks for the detailed explanation, Erick! So the upper instead of exact bound on a targets file size must have been for Uptane, no? @JustinCappos If so, we should document this in an ADR...

**Comment by mnm678** [comment]
In Uptane, the targets file size has to be an exact match so that it is the same for the director and image repos (see step 10 in [full verification](https://github.com/uptane/uptane-standard/blob/master/uptane-standard.md#full-verification-full_verification)).

I think the upper limit should be fine in terms of security, but if no one is using it, we can remove it to eliminate any confusion.

**Comment by lukpueh** [comment]
>  For all metadata files, we may never have the exact size.

Why, @trishankatdatadog? The [spec says](https://github.com/theupdateframework/specification/blame/fd85a8a12403435d84e202ea8790eb4abad4559d/tuf-spec.md#L766):

> LENGTH is the integer length in bytes of the metadata file at METAPATH.

It does say that the value is optional, yes, but to me that does not mean it is *optionally accurate*.  I would expect it to be the exact size if provided.

The main problem here is the parenthesized bit in the client workflow that claims that the number may not be exact. If we remove that there should be no ambiguity about the accuracy of the `length` field, even if the client workflow does a `<=` check. `<=` might just be an artifact of good coding practice.

**Comment by trishankatdatadog** [comment]
> It does say that the value is optional, yes, but to me that does not mean it is _optionally accurate_. I would expect it to be the exact size if provided.

I'm pretty sure we did this for Uptane. Would you please confirm or deny, @pattivacek?

**Comment by pattivacek** [comment]
> > It does say that the value is optional, yes, but to me that does not mean it is _optionally accurate_. I would expect it to be the exact size if provided.
> 
> I'm pretty sure we did this for Uptane. Would you please confirm or deny, @pattivacek?

Uptane removed the requirement to check length and instead relies on version and hash. See for example [here](https://uptane.github.io/uptane-standard/uptane-standard.html#check_targets). IIRC the change was made because the lengths provided no additional security. I think there is some history of the discussion somewhere for the curious.

Note that there is this: "Download up to Z number of bytes [...]. The value for Z is set by the implementer. For example, Z may be tens of kilobytes." However, that doesn't seem relevant here.

**Comment by trishankatdatadog** [comment]
> Uptane removed the requirement to check length and instead relies on version and hash. See for example [here](https://uptane.github.io/uptane-standard/uptane-standard.html#check_targets). IIRC the change was made because the lengths provided no additional security. I think there is some history of the discussion somewhere for the curious.

We don't need to check the length of the targets metadata (which is optional in the snapshot metadata), but we [do](https://uptane.github.io/uptane-standard/uptane-standard.html#verify_image) ask to check the length of a target itself:

> When checking hashes, the ECU SHOULD additionally check that the length of the image matches the length listed in the metadata.

**Comment by pattivacek** [comment]
> We don't need to check the length of the targets metadata (which is optional in the snapshot metadata), but we [do](https://uptane.github.io/uptane-standard/uptane-standard.html#verify_image) ask to check the length of a target itself:
> 
> > When checking hashes, the ECU SHOULD additionally check that the length of the image matches the length listed in the metadata.

Ah yes, sorry. Missed the context there. Anyway, I'm still not aware of a case where you wouldn't know the exact length.

**Comment by trishankatdatadog** [comment]
> Ah yes, sorry. Missed the context there. Anyway, I'm still not aware of a case where you wouldn't know the exact length.

I'm almost certain we added this text for someone's automotive use case. That text sounds so oddly specific to their use case. I'm ok with it, but we just need to decide whether to keep it or not.

**Comment by lukpueh** [comment]
Maybe [theupdateframework/tuf@c5deaa3](https://github.com/theupdateframework/tuf/pull/440/commits/c5deaa340ff33f47fc94e1480240a67639cdbab8) helps to refresh your memory, @trishankatdatadog?

**Comment by trishankatdatadog** [comment]
> Maybe [theupdateframework/tuf@c5deaa3](https://github.com/theupdateframework/tuf/pull/440/commits/c5deaa340ff33f47fc94e1480240a67639cdbab8) helps to refresh your memory, @trishankatdatadog?

But of course, the enemy is always thine own self...

I plead the fifth.

(No, but srsly, does anyone have strong objections about this? That sounds like a valid use case, even if I don't remember exactly who asked for it.)

**Comment by pattivacek** [comment]
> does anyone have strong objections about this? That sounds like a valid use case, even if I don't remember exactly who asked for it.

:) No objections myself, although it seems a bit weird. I don't really get it, but it probably isn't going to break any security guarantees. What I want to know is how you could know the hash but not the length.

**Comment by lukpueh** [comment]
I think I have expressed my strong opinion regarding the seemingly unrelated and apparently unremembered use case, and my not so strong opinion regarding `==` vs. `<=` above. I'm happy to provide a corresponding PR.

**Comment by trishankatdatadog** [comment]
> :) No objections myself, although it seems a bit weird. I don't really get it, but it probably isn't going to break any security guarantees. What I want to know is how you could know the hash but not the length.

Beats me. IIRC this implementor depends completely on this external tool which gives you a root hash but not the length. It is an odd constraint.

**Comment by trishankatdatadog** [comment]
> I think I have expressed my strong opinion regarding the seemingly unrelated and apparently unremembered use case, and my not so strong opinion regarding `==` vs. `<=` above. I'm happy to provide a corresponding PR.

Yes, we certainly need to update either way. If `==`, then we need to say it's exact, not up to. If `<=`, we need to make the length optional in the targets metadata schema. We will also need to check with go-tuf, rust-tuf, tough, etc.

**Comment by mike-sul** [comment]
> What I want to know is how you could know the hash but not the length

e.g. an ostree commit hash, OTA Connect's Target type `OSTREE`?

**Comment by pattivacek** [comment]
> If `<=`, we need to make the length optional in the targets metadata schema. We will also need to check with go-tuf, rust-tuf, tough, etc.

I don't follow. If it's `<=`, that just means the number is inexact, not that it isn't required.

> e.g. an ostree commit hash, OTA Connect's Target type `OSTREE`?

Ha! Good call. Yes, of course, and that's probably where this came from. I was interpreting it differently, but yes, for OSTree we don't know the actual length.

**Comment by lukpueh** [comment]
I agree with @pattivacek that `<=` does not imply that the number is optional. And strictly speaking it doesn't imply that the number is inexact either. All it says is that the match does not have to be exact.

**Comment by trishankatdatadog** [comment]
> Ha! Good call. Yes, of course, and that's probably where this came from. I was interpreting it differently, but yes, for OSTree we don't know the actual length.

Well, it has to be optional for this reason. And if it's optional, we make the check `<=` if it isn't there, and can make it `==` if it is there.

**Comment by trishankatdatadog** [comment]
> e.g. an ostree commit hash, OTA Connect's Target type `OSTREE`?

Thank you, kind stranger. How did you even run into this discussion?

**Comment by lukpueh** [comment]
> we make the check <= if it isn't there

?

**Comment by mike-sul** [comment]
> Thank you, kind stranger. How did you even run into this discussion?

I am "watching" the repo `theupdateframework/specification` and just getting emails on any update in it.

**Comment by trishankatdatadog** [comment]
> ?

!

Do we not agree that an end-user may not always know the exact length?

**Comment by lukpueh** [comment]
> ... we make the check <= if it isn't there ...

To me this reads like `len(target) <= None`, which doesn't make sense.

> Do we not agree that an end-user may not always know the exact length?

Are you talking about "length" in targets metadata being not there or inexact? If so, I would say it is a matter of definition of that "length" field. If we define it to be optional and/or optionally accurate, then I agree that the end-user may not always know the exact length.

**Comment by trishankatdatadog** [comment]
> Are you talking about "length" in targets metadata being not there or inexact? If so, I would say it is a matter of definition of that "length" field. If we define it to be optional and/or optionally accurate, then I agree that the end-user may not always know the exact length.

Oh, I see. Let me be sparkling clear then:

```python
# target = {"hashes": {...}, "length": 42}
length = target.get("length")
if length:
    filepath = download_exactly(url, length)
    assert os.path.getsize(path) == length
else:
    filepath = download_up_to(url, tuf.settings.MAX_TARGET_SIZE)
    assert os.path.getsize(path) <= tuf.settings.MAX_TARGET_SIZE
```

Do we agree?

**Comment by lukpueh** [comment]
Yes, with `MAX_TARGET_SIZE` the `<=` makes more sense. And yes, handling the case where the optional value is missing also makes sense ... under the premise that the value is indeed optional, which so far nothing in the specification suggested.

**Comment by trishankatdatadog** [comment]
> under the premise that the value is indeed optional, which so far nothing in the specification suggested.

It was implicit, and this is what we have to decide whether to make explicit now.

**Comment by mnm678** [comment]
In the case that the exact size isn't known, I think that it makes more sense for the repository to define a max size, as they know more about the likely size of the target. ie:

``` python
length = target.get("length")
filepath = download_up_to(url, length)
assert os.path.getsize(filepath) <= length
```

**Comment by erickt** [comment]
Could someone describe the `ostree` tool some more? I'm still not sure how we'd be in a position where we wouldn't know the download size. Looking at https://ostreedev.github.io/ostree/, it sort of sounds similar to fuchsia packages, where there's a manifest in TUF that points at various blobs, along with the ability to store binary diffs between files, but wouldn't we still know the size of the things we need to fetch?

**Comment by pattivacek** [comment]
> Could someone describe the `ostree` tool some more?

If you'll excuse the promo, try the OSTree section of this page in the [OTA Connect docs](https://docs.ota.here.com/ota-client/latest/comparing-full-filesystem-update-strategies.html). Basically, we don't know the size because you don't know which objects are already on the device and what is missing. You know the commit that you want, but you have to walk the tree and check what you still need to fetch.

> > Thank you, kind stranger. How did you even run into this discussion?
> 
> I am "watching" the repo `theupdateframework/specification` and just getting emails on any update in it.

@mike-sul you're being humble! Mike used to work with me on aktualizr/OTA Connect and is still occasionally involved there.

**Comment by trishankatdatadog** [comment]
> In the case that the exact size isn't known, I think that it makes more sense for the repository to define a max size, as they know more about the likely size of the target. ie:

Actually, this is exactly what my client [does right now](https://github.com/trishankatdatadog/tuf-on-a-plane/blob/4f2b1d5c32a420c8c67197122694583bdf04dfd3/src/tuf_on_a_plane/repository.py#L282-L285), keeps implementation exactly the same between metadata and targets...

**Comment by trishankatdatadog** [comment]
So, before I send any PR, do we have any rough consensus? Should we allow target lengths to be optional? And, if so, checking for target lengths becomes `<=` rather than `==`?

**Comment by pattivacek** [comment]
> Should we allow target lengths to be optional?

Yes, I think so. It should probably be a SHOULD type of thing.

> checking for target lengths becomes `<=` rather than `==`?

I'm not sure these are strictly connected. Would it be better to suggest a maximum target length, particularly if there are not per-target lengths?

**Comment by trishankatdatadog** [comment]
> I'm not sure these are strictly connected. Would it be better to suggest a maximum target length, particularly if there are not per-target lengths?

Sorry, I don't understand. Do you mean the kind of [conditional check](https://github.com/theupdateframework/specification/issues/139#issuecomment-740686354) I proposed earlier in this thread?

**Comment by lukpueh** [comment]
I think the confusion (at least for me) stems from whether you mean `<= length specified in targets metadata`, as suggested by the quoted spec section 5.6.2. and illustrated in Marina's code snippet, or `<= max targets length`, as suggested in your code snippet, or both?

These are the options I see:

1. targets length is mandatory and must be exact
    ```python
    assert len(target_file) == target_info.get("length")
    ```

2. targets length is optional, but must be exact if present
    ```python
    target_len = target_info.get("length")

    if target_len is not None:
      assert len(target_file) == target_len
    else:
      assert len(target_file) <= MAX_TARGET_LENGTH
    ```

3. targets length is optional and may be inexact
    ```python
    assert len(target_file) <= target_info.get("length", MAX_TARGET_LENGTH)
    ```

*NOTE: Above checks are after the fact for brevity and emphasis on the constraints. In reality we should also do `<=` checks  during incremental download to prevent endless data attacks.*

**Comment by trishankatdatadog** [comment]
> These are the options I see:

Thanks for being precise, Lukas. Yes, I meant (2).

> _NOTE: Above checks are after the fact for brevity and emphasis on the constraints. In reality we should also do `<=` checks during incremental download to prevent endless data attacks._

Fine print like this is exactly why we're seemingly talking past each other 😅

**Comment by joshuagl** [comment]
I always assumed the spec says to download "up to the number of bytes" to allow for compression in transit. 

We should never download more than the listed size, but the HTTP server might compress the metadata file before it's transmitted to us. So we might not need to download the full `length` bytes.
